racket/collects/racket/private/list.rkt

```diff
@@ -358,46 +363,58 @@
                ;; need to check this too (see PR 11978)
                (can-compose* name n g f fs)))
       (define (pipeline1 f rfuns)
-        ;; (very) slightly slower alternative:
-        ;; (if (null? rfuns)
-        ;;   f
-        ;;   (pipeline1 (let ([fst (car rfuns)]) (lambda (x) (fst (f x))))
-        ;;              (cdr rfuns)))
-        (lambda (x)
-          (let loop ([x x] [f f] [rfuns rfuns])
-            (if (null? rfuns)
-              (f x)
-              (loop (f x) (car rfuns) (cdr rfuns))))))
+        (cond
+          [(andmap id? rfuns) f]
+          [(id? f) (pipeline1 (car rfuns) (cdr rfuns))]
```

Comment:
This is quadratic time in the case that all functions are identity except the last one in `rfuns`. Maybe special-case an empty `rfuns` after filtering, instead?

Also, as written here, it seems like `f` would not be reduced to single-argument arity like it was before, and maybe also doesn't get renamed as before.